### PR TITLE
Add xilinx network support

### PR DESCRIPTION
**Pull request by mchindri**
Created to debug a new tinyiiod version.
 - For tinyiiod seems to work, but it doesn't has the same performance as in linux.
 - Some file are token from here: https://github.com/Xilinx/embeddedsw/tree/master/lib/sw_apps/lwip_echo_server/src
   because it seems that there is support for multiple xilinx boards.
 - Tested on zc706 and zedboard
 - At the moment lwip must be enabled manually from BSP with LWIP_DHCP set to true. (I failed to add them from makefile but there should be a way. I didn't spend too much time on that )

### CODE REVIEW

**Comment by buha** [comment]
So I just had a look at your branch - you seem to place the xilinx ethernet files here:
```
$ ls ../../network/xil_eth/
i2c_access.c  iic_phyreset.c  platform.c  platform_config.h  platform.h  platform_mb.c  platform_ppc.c  platform_zynq.c  platform_zynqmp.c  sfp.c  si5324.c  xil_eth.c  xil_eth.h
```

By modifying util.tcl like this:
```
-                -template  {Empty Application}
+                -template  {lwIP Echo Server}
```

And runing make on your branch, I get these files:
```
$ ls build/app/src/
echo.c  i2c_access.c  iic_phyreset.c  lscript.ld  main.c  platform.c  platform_config.h  platform.h  platform_mb.c  platform_ppc.c  platform_zynq.c  platform_zynqmp.c  README.txt  sfp.c  si5324.c  Xilinx.spec
```
This shows that the first commit can be dropped, just a slight adaptation of the build system:
 - util.tcl change above + a way to select the template from the Makefile (new parameter for util.tcl?)
 - src.mk get the etnernet files from the autogenerated location
 - remove that build/app/src/main.c (if files are included with SRC_DIRS += ..., otherwise not needed) 
 - edit the build/app/src/platform_config.h file with a script to automatically add a #include "app_config.h" or whatever is needed.
 - ... (maybe other changes)

**Comment by mchindri** [comment]
v2
 - Fix bug (it seems to work ok now)
 - Won't try at this moment to solve the project creation. I will focus on iiod and trigger for the moment.

File: network/xil_eth/xil_eth.c [hunk 1]
```diff
@@ -0,0 +1,654 @@
+/***************************************************************************//**
+ * @file xil_eth.c
+ * @brief Implementation of ethernet functions for network.
+ * @author Mihail Chindris (mihail.chindrs@analog.com)
```
Comment (buha) [inline network/xil_eth/xil_eth.c:4]: some typos in comments:
- your name specifically
- LINUX_SOCKET_H_ i guess is a remnant from a template

**Comment by buha** [comment]
> * Won't try at this moment to solve the project creation. I will focus on iiod and trigger for the moment.

okay, makes sense, we can do that later

**Comment by mchindri** [comment]
v3: 
 - Fix comments
 - Add eth support for xilinx in iio_app (maybe make sens an other PR for it but for the moment it may be ok here)

File: network/xil_eth/xil_eth.c [hunk 2]
```diff
@@ -0,0 +1,654 @@
...
+		/* First buffer. Save it in p */
+		sock->p = p;
+		sock->p_idx = 0;
+	} else {
+		/*
+		 * TODO: Can this cause error while reading?
+		 * Should some locking be use here?
+		 */
+		/* Concatenate new buffer to p */
+		pbuf_cat(sock->p, p);
```
Comment (adrimbarean) [inline network/xil_eth/xil_eth.c:325]: Question: do we need to update `sock->p_idx` here or does `pbuf_cat` update it itself?

File: network/xil_eth/xil_eth.c [hunk 3]
```diff
@@ -0,0 +1,654 @@
...
+
+	if (serv_sock->state != SOCKET_ACCEPTING) {
+		if (serv_sock->state != SOCKET_LISTENING)
+			return -EINVAL;
+		tcp_accept(serv_sock->pcb, xil_eth_accept_callback);
+		serv_sock->state = SOCKET_ACCEPTING;
+	}
+
+	for (i = 0; i < MAX_SOCKETS; ++i) {
+		cli_sock = &desc->sockets[i];
```
Comment (adrimbarean) [inline network/xil_eth/xil_eth.c:575]: nitpick: why not `_get_sock` here as well?

File: iio/iio_app/iio_app.c [hunk 4]
```diff
@@ -47,36 +47,53 @@
 #include "no-os/delay.h"
 #include <stdio.h>
 
+#undef USE_TCP_SOCKET
```
Comment (adrimbarean) [inline iio/iio_app/iio_app.c:50]: Question: what is the point of undefining `USE_TCP_SOCKET` here?

File: network/xil_eth/xil_eth.c (hunk 2 above)
Comment (mchindri) [reply network/xil_eth/xil_eth.c:325]: I assume p is put at the end of the list. 
p_idx is the index in the buffer of the first element

File: iio/iio_app/iio_app.c (hunk 4 above)
Comment (mchindri) [reply iio/iio_app/iio_app.c:50]: I forgot to remove it. It was used just for testing

File: network/xil_eth/xil_eth.c [hunk 5]
```diff
@@ -0,0 +1,654 @@
...
+		/* DHCP Config */
+
+		/* Create a new DHCP client for this interface.
+		* Note: you must call dhcp_fine_tmr() and dhcp_coarse_tmr() at
+		* the predefined regular intervals after starting the client.
+		*/
+		dhcp_start(netif);
+		dhcp_timoutcntr = 24;
+
+		while(((netif->ip_addr.addr) == 0) && (dhcp_timoutcntr > 0))
```
Comment (buha) [inline network/xil_eth/xil_eth.c:220]: did you mean `dhcp_timoutcntr--`?

File: network/xil_eth/xil_eth.c [hunk 6]
```diff
@@ -0,0 +1,654 @@
...
+			xemacif_input(netif);
+
+		if (dhcp_timoutcntr <= 0) {
+			if ((netif->ip_addr.addr) == 0) {
+				pr_err("DHCP Timeout\r\n");
+				ret = FAILURE;
```
Comment (buha) [inline network/xil_eth/xil_eth.c:226]: I think there is an -ETIMEDOUT

File: network/xil_eth/xil_eth.c [hunk 7]
```diff
@@ -0,0 +1,654 @@
...
+
+	netif = &ldesc->netif;
+	/* Needed by platform files */
+	echo_netif = netif;
+
+	/* Add network interface to the netif_list, and set it as default */
+	if (!xemac_add(netif, &ipaddr, &netmask, &gw,
+		       param->mac_ethernet_address, param->emac_baseaddr)) {
+		pr_err("Error adding N/W interface\n\r");
+		ret = FAILURE;
```
Comment (buha) [inline network/xil_eth/xil_eth.c:198]: also here, a more meaningful error code

File: network/xil_eth/xil_eth.c [hunk 8]
```diff
@@ -0,0 +1,654 @@
...
+	}
+	xemacif_input(&desc->netif);
+
+	return SUCCESS;
+}
+
+int32_t xil_eth_get_network_interface(struct xil_eth_desc *desc,
+				      struct network_interface **net)
+{
+	*net = &desc->noos_net;
```
Comment (buha) [inline network/xil_eth/xil_eth.c:276]: this dereferencing could fail

**Pull request closed without merging**